File: src/main/java/com/ptaf/ai/quality/QualityReportWriter.java
package com.ptaf.ai.quality;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;

/** Writes {@code target/ai-quality-report.json} and {@code target/ai-quality-report.html}. */
public final class QualityReportWriter {

    private static final ObjectMapper JSON = new ObjectMapper();

    private QualityReportWriter() {
    }

    public static void write(Path targetDir, QualityReport report) throws Exception {
        Files.createDirectories(targetDir);
        Path jsonPath = targetDir.resolve("ai-quality-report.json");
        Path htmlPath = targetDir.resolve("ai-quality-report.html");

        ObjectNode root = JSON.createObjectNode();
        root.put("generatedAt", Instant.now().toString());
        root.put("warnThreshold", report.warnThreshold());
        root.put("failedStrict", report.failedStrict());
        ArrayNode syn = root.putArray("syntaxIssues");
        for (String s : report.syntaxIssues()) {
            syn.add(s);
        }
        ArrayNode dups = root.putArray("duplicateStepGroups");
        for (DuplicateStepDetector.DuplicateGroup g : report.duplicateGroups()) {
            ObjectNode go = dups.addObject();
            go.put("normalized", g.normalizedStep());
            go.put("count", g.count());
            ArrayNode occ = go.putArray("occurrences");
            for (StepOccurrence o : g.occurrences()) {
                ObjectNode oo = occ.addObject();
                oo.put("file", o.relativePath());
                oo.put("line", o.lineNumber());
                oo.put("raw", o.rawLine());
            }
        }
        Files.writeString(jsonPath, JSON.writerWithDefaultPrettyPrinter().writeValueAsString(root), StandardCharsets.UTF_8);

        StringBuilder html = new StringBuilder();
        html.append("<!DOCTYPE html><html><head><meta charset=\"utf-8\"><title>PTAF AI quality</title></head><body>\n");
        html.append("<h1>AI quality report</h1><p>").append(Instant.now()).append("</p>\n");
        html.append("<h2>Syntax / structure</h2><ul>\n");
        for (String s : report.syntaxIssues()) {
            html.append("<li>").append(escapeHtml(s)).append("</li>\n");
        }
        html.append("</ul>\n<h2>Duplicate step groups (≥ ").append(report.warnThreshold()).append(")</h2>\n");
        for (DuplicateStepDetector.DuplicateGroup g : report.duplicateGroups()) {
            html.append("<h3>").append(escapeHtml(g.normalizedStep())).append(" (").append(g.count()).append(")</h3><ul>\n");
            for (StepOccurrence o : g.occurrences()) {
                html.append("<li>").append(escapeHtml(o.relativePath())).append(":").append(o.lineNumber()).append("</li>\n");
            }
            html.append("</ul>\n");
        }
        html.append("</body></html>");
        Files.writeString(htmlPath, html.toString(), StandardCharsets.UTF_8);
    }

    private static String escapeHtml(String s) {
        if (s == null) {
            return "";
        }
        return s.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;").replace("\"", "&quot;");
    }
}

File: src/main/ptaf/ai/quality/quality_report_writer.py
from __future__ import annotations

import html
import json
from dataclasses import asdict
from datetime import datetime, timezone
from pathlib import Path

from ptaf.ai.quality.quality_gate_service import QualityReport


def write(target_dir: Path, report: QualityReport) -> None:
    target_dir.mkdir(parents=True, exist_ok=True)
    json_path = target_dir / "ai-quality-report.json"
    html_path = target_dir / "ai-quality-report.html"

    payload = {
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "warnThreshold": report.warn_threshold,
        "failedStrict": report.failed_strict,
        "syntaxIssues": report.syntax_issues,
        "duplicateStepGroups": [
            {
                "normalized": group.normalized_step,
                "count": group.count,
                "occurrences": [asdict(occ) for occ in group.occurrences],
            }
            for group in report.duplicate_groups
        ],
    }
    json_path.write_text(json.dumps(payload, indent=2), encoding="utf-8")

    lines = [
        "<!DOCTYPE html><html><head><meta charset=\"utf-8\"><title>PTAF AI quality</title></head><body>",
        f"<h1>AI quality report</h1><p>{datetime.now(timezone.utc).isoformat()}</p>",
        "<h2>Syntax / structure</h2><ul>",
    ]
    for issue in report.syntax_issues:
        lines.append(f"<li>{html.escape(issue)}</li>")
    lines.append(f"</ul><h2>Duplicate step groups (≥ {report.warn_threshold})</h2>")
    for group in report.duplicate_groups:
        lines.append(
            f"<h3>{html.escape(group.normalized_step)} ({group.count})</h3><ul>"
        )
        for occurrence in group.occurrences:
            lines.append(
                f"<li>{html.escape(occurrence.relative_path)}:{occurrence.line_number}</li>"
            )
        lines.append("</ul>")
    lines.append("</body></html>")
    html_path.write_text("\n".join(lines), encoding="utf-8")
